### ISSUE TITLE
Incorrect "dereference.of.nullable" when final variable is checked for null and then dereferenced in anonymous class

### ISSUE THREAD

**Issue by sebright** [typetools/checker-framework#2011]
I reproduced this issue in a small example with the live demo:

```java
import org.checkerframework.checker.nullness.qual.Nullable;
public class MyClass {
  @Nullable public Runnable toStringRunnable(@Nullable final Object arg) {
    if (arg == null) return null;
    return new Runnable() {
      @Override
      public void run() {
        arg.toString();
      }
    };
  }
}
```

No. | Type | Description | Line | Column
-- | -- | -- | -- | --
1 | error | Error: [dereference.of.nullable] dereference of possibly-null reference arg | 8 | 9

Link to demo: http://eisop.uwaterloo.ca/live/#mode=display&code=import+org.checkerframework.checker.nullness.qual.Nullable%3B%0Apublic+class+MyClass+%7B%0A++%40Nullable+public+Runnable+toStringRunnable(%40Nullable+final+Object+arg)+%7B%0A++++if+(arg+%3D%3D+null)+return+null%3B%0A++++return+new+Runnable()+%7B%0A++++++%40Override%0A++++++public+void+run()+%7B%0A++++++++arg.toString()%3B%0A++++++%7D%0A++++%7D%3B%0A++%7D%0A%7D&typeSystem=nullness

I expected `arg` to be non-null where `toString()` is called, because `arg` is final, and it is checked for null before the Runnable is created.  If I move the statement `arg.toString();` to the line before the Runnable is created, the null analysis passes.

**Comment by wmdietl** [comment]
This is a duplicate of #266.
For the currently-suggested work-around see https://github.com/typetools/checker-framework/issues/266#issuecomment-118318478

I've been looking at a fix for #266 lately, but it is unfortunately not yet ready.
